File: javabackend/src/main/java/mft/dev/javabackend/service/implementation/TodoServiceImpl.java
package mft.dev.javabackend.service.implementation;

import jakarta.transaction.Transactional;
import mft.dev.javabackend.dto.insert.request.TodoInsertRequestDTO;
import mft.dev.javabackend.dto.update.TodoUpdateRequestDTO;
import mft.dev.javabackend.model.Todo;
import mft.dev.javabackend.model.Utente;
import mft.dev.javabackend.repository.TodoRepository;
import mft.dev.javabackend.service.TodoService;
import mft.dev.javabackend.service.UtenteService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class TodoServiceImpl implements TodoService {

    @Autowired
    private TodoRepository todoRepository;

    @Autowired
    private UtenteService utenteService;

    @Override
    public List<Todo> getTodosByUtenteId(long id) {
        Utente utenteFound = this.utenteService.getUtente(id);

        return this.todoRepository.findByUtente(utenteFound);
    }

    @Override
    public Todo getTodo(long id) {
        return this.todoRepository.findById(id).orElseThrow();
    }

    @Override
    @Transactional
    public Todo insertTodo(long id, TodoInsertRequestDTO dto) {
        Utente utenteFound = this.utenteService.getUtente(id);

        Todo newTodo = dto.toTodo(utenteFound);

        return this.todoRepository.save(newTodo);
    }

    @Override
    @Transactional
    public List<Todo> insertMoltiTodo(long id, List<TodoInsertRequestDTO> dto) {
        Utente utenteFound = this.utenteService.getUtente(id);

        List<Todo> newTodos = dto.stream()
                .map(t -> t.toTodo(utenteFound))
                .toList();

        return this.todoRepository.saveAll(newTodos);
    }

    @Override
    @Transactional
    public Todo updateTodo(long id, TodoUpdateRequestDTO dto) {
        Todo todoFound = this.getTodo(id);

        todoFound = TodoUpdateRequestDTO.toTodo(dto, todoFound);

        return this.todoRepository.save(todoFound);
    }

    @Override
    @Transactional
    public void deleteTodo(long id) {
        Todo todoFound = this.getTodo(id);

        this.todoRepository.delete(todoFound);
    }

    @Override
    @Transactional
    public void deleteAllTodoByUtenteId(long id) {
        Utente utente = this.utenteService.getUtente(id);

        this.todoRepository.deleteAllByUtente(utente);
    }
}

File: kotlinbackend/src/main/kotlin/mft/dev/kotlinbackend/service/implementation/TodoServiceImpl.kt
package mft.dev.kotlinbackend.service.implementation

import mft.dev.kotlinbackend.dto.insert.TodoInsertRequestDTO
import mft.dev.kotlinbackend.dto.update.TodoUpdateRequestDTO
import mft.dev.kotlinbackend.model.Todo
import mft.dev.kotlinbackend.model.Utente
import mft.dev.kotlinbackend.repository.TodoRepository
import mft.dev.kotlinbackend.service.TodoService
import mft.dev.kotlinbackend.service.UtenteService
import org.springframework.stereotype.Service

@Service
class TodoServiceImpl(
    private val todoRepository: TodoRepository,
    private val utenteService: UtenteService
) : TodoService {
    override fun getTodosByUtenteId(id: Long): MutableList<Todo> {
        val utente: Utente = this.utenteService.getUtente(id)

        return this.todoRepository.findByUtente(utente)
    }

    override fun getTodo(id: Long): Todo {
        return this.todoRepository.findById(id).orElseThrow()
    }

    override fun insertTodo(id: Long, dto: TodoInsertRequestDTO): Todo {
        val utenteFound: Utente = this.utenteService.getUtente(id)

        val todo: Todo = dto.toTodo(utenteFound)

        return this.todoRepository.save(todo)
    }

    override fun insertMoltiTodo(id: Long, dto: MutableList<TodoInsertRequestDTO>): MutableList<Todo> {
        val utenteFound: Utente = this.utenteService.getUtente(id)

        val todos: MutableList<Todo> = dto.map { it.toTodo(utenteFound) }.toMutableList()

        return this.todoRepository.saveAll(todos)
    }

    override fun updateTodo(id: Long, dto: TodoUpdateRequestDTO): Todo {
        val todoFound: Todo = this.getTodo(id)

        TodoUpdateRequestDTO.toTodo(dto, todoFound)

        return todoFound
    }

    override fun deleteTodo(id: Long) {
        val todoFound: Todo = this.getTodo(id)

        this.todoRepository.delete(todoFound)
    }

    override fun deleteAllTodoByUtenteId(id: Long) {
        val utenteFound: Utente = this.utenteService.getUtente(id)

        this.todoRepository.deleteByUtente(utenteFound)
    }
}